experimental/displaynames/src/displaynames.rs

```diff
@@ -399,90 +407,209 @@ impl LocaleDisplayNamesFormatter {
     );
 
     /// Returns the display name of a locale.
+    /// This implementation is based on the algorithm described in
+    /// https://www.unicode.org/reports/tr35/tr35-general.html#locale_display_name_algorithm
+    /// 
     // TODO: Make this return a writeable instead of using alloc
     pub fn of<'a, 'b: 'a, 'c: 'a>(&'b self, locale: &'c Locale) -> Cow<'a, str> {
-        // https://www.unicode.org/reports/tr35/tr35-general.html#Display_Name_Elements
+        // Step - 1: Construct the canonical locale from a given locale.
+        let cannonical_locale = Locale::canonicalize(locale.to_string()).unwrap();
+
+        let longest_prefix =
+            match find_longest_matching_prefix(&cannonical_locale, &self.locale_data.get().names) {
+                Some(prefix) => prefix,
+                None => cannonical_locale
+                    .split_once("-")
+                    .unwrap_or((&cannonical_locale, ""))
+                    .0
+                    .to_string(),
+            };
+
+        // Step - 2: Construct a locale display name string (LDN).
+        let ldn = get_locale_display_name(&longest_prefix, &self);
+
+        // Step - 3: Construct a vector of longest qualifying substrings (LQS).
+        let lqs = get_longest_qualifying_substrings(&cannonical_locale, &longest_prefix, &self);
+
+        // Step - 4: Return the displayname based on the size of LQS.
+        if lqs.len() == 0 {
+            return ldn.to_string().into();
```

Comment:
This can probably return a borrowed string? I struggled  with returning strings from the function while implementing this PR. Would appreciate any feedback.